### PR TITLE
feat: complete `eval_expression` and `eval_module` support

### PR DESCRIPTION
**Pull request by AayushSabharwal**
## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
https://github.com/SciML/ModelingToolkit.jl/actions/runs/9714593463/job/26814397836?pr=2826#step:6:1871 the failures look real

**Comment by AayushSabharwal** [comment]
The inference failures are weird, even Cthulhu output doesn't make sense.

File: src/systems/abstractsystem.jl [hunk 1]
```diff
@@ -1210,19 +1223,30 @@ end
...
+
+# This is hit because ensemble problems do a deepcopy
+function Base.deepcopy_internal(ofc::ObservedFunctionCache, stackdict::IdDict)
+    sys = deepcopy(ofc.sys)
+    dict = deepcopy(ofc.dict)
+    eval_expression = ofc.eval_expression
+    eval_module = ofc.eval_module
+    newofc = ObservedFunctionCache(sys, dict, eval_expression, eval_module)
+    stackdict[ofc] = newofc
+    return newofc
```
Comment (ChrisRackauckas) [inline src/systems/abstractsystem.jl:1242]: This is not a good idea, it's not a public API. What's the real problem here?

Comment (AayushSabharwal) [reply src/systems/abstractsystem.jl:1242]: It's in `deepcopy`'s docstring to overload this for custom `deepcopy`. The problem here is that ensemble simulations deepcopy the problem, which tries to deepcopy the module inside `ObservedFunctionCache`

**Pull request merged**